File: src/main/java/com/xu/algorithm/L5LongestPalindromicSubstring.java
package com.xu.algorithm;

public class L5LongestPalindromicSubstring {

    public String longestPalindrome(String s) {
        int length = s.length();
        if (length < 2) return s;
        Boolean[][] tmp = new Boolean[length][length];
        int begin = 0;
        int end = 0;
        char[] chars = s.toCharArray();

        for (int i = length - 1; i >= 0; i--) {
            for (int j = i; j < length; j++) {
                if (chars[i] != chars[j]) {
                    tmp[i][j] = false;
                } else {
                    if (j - i < 3) {
                        tmp[i][j] = true;
                    }else {
                        tmp[i][j] = tmp[i+1][j-1];
                    }
                }
                if(tmp[i][j] && end-begin<=j-i){
                    begin=i;
                    end=j;
                }
            }
        }


        return s.substring(begin,end+1);
    }


    public static void main(String[] args) {
        String s = new L5LongestPalindromicSubstring().longestPalindrome("ac");
        System.out.println(s);
    }
}

File: src/main/scala/com/xu/algorithm/L5LongestPalindromicSubstring.scala
//package com.xu.algorithm
//
//
///**
// * 最大回文子串
// */
//class L5LongestPalindromicSubstring {
//
//  def longestPalindrome(s: String): String = {
//
//    if (s.trim.isEmpty) return ""
//    val dpTmp = Array.ofDim[Boolean](s.length, s.length)
//    for (i <- s.length to 0) {
//      for (j <- i to s.length) {
//        if (s.charAt(i) == s.charAt(j)) {
//          if (i - 1 >= j + 1) {
//            if (dpTmp(i + 1)(j - 1)) {
//              dpTmp(i)(j) = true
//            }
//          }else dpTmp(i)(j)=true
//        }
//        if(dpTmp(i)(j)){
//
//        }
//      }
//    }
//
//
//    ""
//  }
//
//}
//
//
//object L5LongestPalindromicSubstring {
//
//
//  def main(args: Array[String]): Unit = {
//    new L5LongestPalindromicSubstring().longestPalindrome("")
//  }
//
//}